modal_proto/api.proto

```diff
@@ -720,6 +743,86 @@ message ContainerExecWaitResponse {
   bool completed = 2;
 }
 
+message ContainerFileCloseRequest {
+  string file_descriptor = 1;
+}
+
+message ContainerFileDeleteBytesRequest {
+  string file_descriptor = 1;
+  optional uint32 start_inclusive = 2;
+  optional uint32 end_exclusive = 3;
+}
+
+message ContainerFileFlushRequest {
+  string file_descriptor = 1;
+}
+
+message ContainerFileOpenRequest {
+  // file descriptor is hydrated when sent from server -> worker
+  optional string file_descriptor = 1;
+  string path = 2;
+  string mode = 3;
+}
+
+message ContainerFileReadLineRequest {
+  string file_descriptor = 1;
+}
+
+message ContainerFileReadRequest {
+  string file_descriptor = 1; 
+  optional uint32 n = 2;
+}
+
+message ContainerFileSeekRequest {
+  string file_descriptor = 1;
+  int32 offset = 2;
+  SeekWhence whence = 3;
+}
+
+message ContainerFileWriteReplaceBytesRequest {
+  string file_descriptor = 1;
+  bytes data = 2;
+  optional uint32 start_inclusive = 3;
+  optional uint32 end_exclusive = 4;
+}
+
+message ContainerFileWriteRequest {
+  string file_descriptor = 1;
+  bytes data = 2;
+}
+
+message ContainerFilesystemExecGetOutputRequest {
+  string exec_id = 1;
+  float timeout = 2;
+  uint64 last_batch_index = 3;
+}
+
+message ContainerFilesystemExecRequest {
+  oneof file_exec_request_oneof {
+    ContainerFileOpenRequest file_open_request = 1;
+    ContainerFileWriteRequest file_write_request = 2;
+    ContainerFileReadRequest file_read_request = 3;
+    ContainerFileFlushRequest file_flush_request = 4;
+    ContainerFileReadLineRequest file_read_line_request = 5;
+    ContainerFileSeekRequest file_seek_request = 6;
+    ContainerFileDeleteBytesRequest file_delete_bytes_request = 7;
+    ContainerFileWriteReplaceBytesRequest file_write_replace_bytes_request = 8;
+    ContainerFileCloseRequest file_close_request = 9;
+  }
+  string task_id = 10;
+}
+
+message ContainerFilesystemExecResponse {
+  string exec_id = 1;
+  // only set when the request opens a new file, i.e., ContainerFileOpenRequest
+  optional string file_descriptor = 2;
+}
+
+message ContainerFilesystemGetOutputRequest {
+  float timeout = 1;
+  uint64 last_batch_index = 2;
```

Comment:
It's actually a mistake that the current RPCs have last_batch_index - it doesn't work with `DataChannel`s. Let's remove it. We should keep `batch_index` on the batch responses just for informational reasons though.